### PR TITLE
[SPARK-29227][SS] Track rule info in optimization phase

### PR DESCRIPTION
**Pull request by wenxuanguan**
### What changes were proposed in this pull request?

Track timing info for each rule in optimization phase using `QueryPlanningTracker` in Structured Streaming

### Why are the changes needed?

In Structured Streaming we only track rule info in analysis phase, not in optimization phase.

### Does this PR introduce any user-facing change?

No

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java [hunk 1]
```diff
@@ -29,7 +29,7 @@
   /**
    * Delete data from a data source table that matches filter expressions.
    * <p>
-   * Rows are deleted from the data source iff all of the filter expressions match. That is, the
+   * Rows are deleted from the data source if all of the filter expressions match. That is, the
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: iff isn't a typo. it's "if and only if"

Comment (wenxuanguan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsDelete.java]: reverted

**Comment by gaborgsomogyi** [comment]
Just for my own understanding where such tracking info can be seen after it's collected?

**Comment by wenxuanguan** [comment]
> Just for my own understanding where such tracking info can be seen after it's collected?

@gaborgsomogyi Thanks for your reply.
Follow by jira [SPARK-26221](https://issues.apache.org/jira/browse/SPARK-26221), it is improvements for better metrics and instrumentation.  And the reporting functions can be used for self-test at present.

**Review by HeartSaVioR — APPROVED** [review]
I see what's missing here.

https://github.com/apache/spark/blob/a1b90bfc0faa2b4b2b7388443a734a217792d585/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala#L91-L110

For now elapsed time for each rule would not be measured in streaming query, and this patch fixes it.

In batch query it does correctly, see below:
https://github.com/apache/spark/blob/a1b90bfc0faa2b4b2b7388443a734a217792d585/sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala#L77-L81

LGTM

**Review by gaborgsomogyi — APPROVED** [review]
LGTM. Checked a query output manually.
Haven't had a super deep look but I think it can be unit tested.

**Comment by wenxuanguan** [comment]
@HyukjinKwon  @dongjoon-hyun  Is there any thing I can do to move on, Thanks.

**Comment by HyukjinKwon** [comment]
ok to test

**Review by HyukjinKwon — COMMENTED** [review]
Yes, can we add a test? (see https://github.com/apache/spark/pull/23096)

**Comment by wenxuanguan** [comment]
Add commit for UT. 
The only change is in `QueryPlanningTrackerEndToEndSuite`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala [hunk 2]
```diff
@@ -38,4 +40,22 @@ class QueryPlanningTrackerEndToEndSuite extends SharedSparkSession {
     assert(tracker.rules.nonEmpty)
   }
 
+  test("streaming") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala]:
```suggestion
  test("SPARK-29227: Track rule info in optimization phase in streaming") {
```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by gaborgsomogyi** [comment]
retest this please

**Review by HeartSaVioR — APPROVED** [review]
LGTM again

**Comment by wenxuanguan** [comment]
merge master to redo check, nothing changed.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala [hunk 3]
```diff
@@ -38,4 +40,22 @@ class QueryPlanningTrackerEndToEndSuite extends SharedSparkSession {
     assert(tracker.rules.nonEmpty)
   }
 
+  test("SPARK-29227: Track rule info in optimization phase in streaming") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala]: For the new feature, we don't use Apache JIRA issue prefix. Please remove `SPARK-29227: `.

Comment (wenxuanguan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala]: @dongjoon-hyun Thanks for your reply.
I fell confused about test name if just delele `SPARK-29227: `, cause this ut is to test tracker phase and rule info in streaming, not only rule info in optimization phase as jira title described. So should I delete `SPARK-29227: Track rule info in optimization phase in `?
And also, do you mean add Apache JIRA issue info(prefix and title) in test name if it is bugfix?

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala]: Let me just merge it since it's rather a nit.

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala]: No, @wenxuanguan ~ Only the JIRA ID.
> And also, do you mean add Apache JIRA issue info(prefix and title) in test name if it is bugfix?

Comment (wenxuanguan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/QueryPlanningTrackerEndToEndSuite.scala]: > No, @wenxuanguan ~ Only the JIRA ID.
> 
> > And also, do you mean add Apache JIRA issue info(prefix and title) in test name if it is bugfix?

@dongjoon-hyun got it, Thanks a lot.

**Comment by wenxuanguan** [comment]
@HeartSaVioR @gaborgsomogyi @HyukjinKwon @dongjoon-hyun Thank you all for review and merge.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 4]
```diff
@@ -77,7 +77,8 @@ class IncrementalExecution(
    */
   override
   lazy val optimizedPlan: LogicalPlan = tracker.measurePhase(QueryPlanningTracker.OPTIMIZATION) {
-    sparkSession.sessionState.optimizer.execute(withCachedData) transformAllExpressions {
+    sparkSession.sessionState.optimizer.executeAndTrack(withCachedData,
+      tracker) transformAllExpressions {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala]: nitpit: in most cases, we do not break the function call in the middle of parameter lists. We can change it to 
```Scala
    val sessionState = sparkSession.sessionState
    sessionState.optimizer.executeAndTrack(withCachedData, tracker).transformAllExpressions {
```
or 
```Scala
    sparkSession.sessionState.optimizer
      .executeAndTrack(withCachedData, tracker).transformAllExpressions {
```

**Pull request closed without merging**